app/models/custom_hint.rb

```diff
@@ -0,0 +1,87 @@
+# Loads our handcrafted hints. We collaborate on these in a google sheet, but
+# we draw them from a separate CSV file rather than automatically from Sheets
+# for two reasons:
+#  1) to avoid having a production app have reliance on a developer's personal
+#     API key
+#  2) because the content of the google sheet is not guaranteed to be
+#     deployable - librarians use it as an active workspace - so we want them
+#     to take a specific action that indicates readiness before we ingest.
+#
+require 'csv'
+require 'open-uri'
+class CustomHint
+  attr_reader :url
+
+  def initialize(url)
+    @url = url
+    # Validate the URL ~before~ setting @csv; there's no point in trying to
+    # fetch content over the network if we know the URL is bogus.
+    process_url
+    @csv = csv
+    @source = 'custom'
+  end
+
+  def process_url
+    canonicalize_url
+    validate_url
+  end
+
+  def validate_url
+    as_uri = URI(@url)
+
+    raise "Invalid URL - not a Dropbox download URL" unless [
```

Comment:
Ensuring that it ends in `dl=1` is dropbox-specific; in general I can be reasonably confident that the thing I'm getting served is *a file* and not *an html page* (that I would have to parse in a different way) by doing this, whereas if I were allowing more general URLs I'd want to be more paranoid about validation.